File: sdk/go/worker_trust_build.go
package capsdk

import (
	"bytes"
	"fmt"
	"strings"
	"time"

	agentv1 "github.com/cordum-io/cap/v2/cordum/agent/v1"
	"google.golang.org/protobuf/proto"
	"google.golang.org/protobuf/types/known/timestamppb"
)

// VerifiedWorkerHandshakeChallenge is an opaque, scheduler-authenticated
// challenge. Its contents can only be obtained as a defensive copy.
type VerifiedWorkerHandshakeChallenge struct {
	request  *agentv1.BusPacket
	response *agentv1.BusPacket
}

// Message returns a copy safe for embedding in an authenticate packet.
func (v *VerifiedWorkerHandshakeChallenge) Message() *agentv1.WorkerHandshakeChallenge {
	if v == nil || v.response == nil || v.response.GetWorkerHandshakeChallenge() == nil {
		return nil
	}
	return proto.Clone(v.response.GetWorkerHandshakeChallenge()).(*agentv1.WorkerHandshakeChallenge)
}

// BuildWorkerHandshakeChallengeRequest creates and signs one protobuf request.
func BuildWorkerHandshakeChallengeRequest(config *WorkerTrustConfig, options WorkerHandshakeRequestOptions) (*agentv1.BusPacket, error) {
	if err := ValidateWorkerTrustConfig(config); err != nil {
		return nil, err
	}
	if err := validateRequestOptions(options); err != nil {
		return nil, err
	}
	request := &agentv1.WorkerHandshakeChallengeRequest{
		RequestId: options.RequestID, TraceId: options.TraceID, WorkerId: config.WorkerID,
		ProofKeyId: config.ProofKeyID, ProofAlgorithm: p256WorkerProofAlgorithm(), Audience: config.Audience,
		Purpose: options.Purpose, ClientNonce: append([]byte(nil), options.ClientNonce...),
		ProtocolVersion: WorkerHandshakeProtocolVersion, SdkVersion: config.SDKVersion,
	}
	packet := newWorkerTrustPacket(options.TraceID, config.WorkerID, options.CreatedAt)
	packet.Payload = &agentv1.BusPacket_WorkerHandshakeChallengeRequest{WorkerHandshakeChallengeRequest: request}
	if err := SignTrustHandshake(packet, config.ProofPrivateKey); err != nil {
		return nil, fmt.Errorf("build worker handshake challenge request: %w", err)
	}
	if err := ValidateWorkerTrustPacket(packet); err != nil {
		return nil, fmt.Errorf("%w: %w", ErrWorkerHandshakePacket, err)
	}
	return packet, nil
}

// BuildWorkerHandshakeAuthenticate signs the verified challenge and capability.
func BuildWorkerHandshakeAuthenticate(config *WorkerTrustConfig, verified *VerifiedWorkerHandshakeChallenge, capability *agentv1.Handshake, currentSessionToken string, createdAt time.Time) (*agentv1.BusPacket, error) {
	if err := ValidateWorkerTrustConfig(config); err != nil {
		return nil, err
	}
	challenge := verified.Message()
	if err := validateVerifiedChallengeConfig(config, challenge); err != nil {
		return nil, err
	}
	if err := validateWorkerCapability(capability, challenge); err != nil {
		return nil, err
	}
	if err := validateSessionForPurpose(challenge.GetPurpose(), currentSessionToken); err != nil {
		return nil, err
	}
	if createdAt.IsZero() {
		return nil, fmt.Errorf("%w: created_at is empty", ErrWorkerHandshakePacket)
	}
	packet := newWorkerTrustPacket(challenge.GetTraceId(), config.WorkerID, createdAt)
	packet.AuthToken = currentSessionToken
	packet.Payload = &agentv1.BusPacket_WorkerHandshakeAuthenticate{WorkerHandshakeAuthenticate: &agentv1.WorkerHandshakeAuthenticate{
		Challenge: challenge, CapabilityHandshake: proto.Clone(capability).(*agentv1.Handshake),
	}}
	if err := SignTrustHandshake(packet, config.ProofPrivateKey); err != nil {
		return nil, fmt.Errorf("build worker handshake authenticate: %w", err)
	}
	if err := ValidateWorkerTrustPacket(packet); err != nil {
		return nil, fmt.Errorf("%w: %w", ErrWorkerHandshakePacket, err)
	}
	return packet, nil
}

func validateRequestOptions(options WorkerHandshakeRequestOptions) error {
	if !canonicalTrustString(options.RequestID) || !canonicalTrustString(options.TraceID) {
		return fmt.Errorf("%w: request_id and trace_id are required", ErrWorkerHandshakePacket)
	}
	if !validWorkerHandshakePurpose(options.Purpose) {
		return fmt.Errorf("%w: unsupported purpose %d", ErrWorkerHandshakePacket, options.Purpose)
	}
	if len(options.ClientNonce) != WorkerHandshakeNonceSize {
		return fmt.Errorf("%w: client nonce must be %d bytes", ErrWorkerHandshakePacket, WorkerHandshakeNonceSize)
	}
	if options.CreatedAt.IsZero() {
		return fmt.Errorf("%w: created_at is empty", ErrWorkerHandshakePacket)
	}
	return nil
}

func newWorkerTrustPacket(traceID, senderID string, createdAt time.Time) *agentv1.BusPacket {
	return &agentv1.BusPacket{
		TraceId: traceID, SenderId: senderID, CreatedAt: timestamppb.New(createdAt.UTC()),
		ProtocolVersion: WorkerHandshakeProtocolVersion,
	}
}

func validateVerifiedChallengeConfig(config *WorkerTrustConfig, challenge *agentv1.WorkerHandshakeChallenge) error {
	if challenge == nil {
		return fmt.Errorf("%w: challenge is unverified", ErrWorkerHandshakeBinding)
	}
	if challenge.GetWorkerId() != config.WorkerID || challenge.GetAgentId() != config.ExpectedAgentID ||
		challenge.GetTenantId() != config.TenantID || challenge.GetProofKeyId() != config.ProofKeyID ||
		challenge.GetAudience() != config.Audience || challenge.GetSdkVersion() != config.SDKVersion {
		return fmt.Errorf("%w: challenge identity changed", ErrWorkerHandshakeBinding)
	}
	return nil
}

func validateSessionForPurpose(purpose agentv1.WorkerHandshakePurpose, token string) error {
	if purpose == issueWorkerHandshakePurpose() && token != "" {
		return fmt.Errorf("%w: issue must not carry a session", ErrWorkerHandshakePacket)
	}
	if purpose == renewWorkerHandshakePurpose() && !validSessionToken(token) {
		return fmt.Errorf("%w: renew requires the current session", ErrWorkerHandshakePacket)
	}
	return nil
}

func validSessionToken(token string) bool {
	if token == "" || token != strings.TrimSpace(token) || len(token) > WorkerHandshakeMaxSessionTokenSize {
		return false
	}
	for index := 0; index < len(token); index++ {
		if token[index] < 0x21 || token[index] > 0x7e {
			return false
		}
	}
	return true
}

func sameNonce(left, right []byte) bool {
	return bytes.Equal(left, right)
}

File: sdk/python/cap/worker_trust_build.py
"""Builders for signed worker trust-handshake requests."""

from datetime import datetime
from typing import Optional

from cap.pb.cordum.agent.v1 import buspacket_pb2, handshake_pb2
from cap.trust_signing import sign_trust_packet
from cap.worker_trust import (
    WORKER_HANDSHAKE_NONCE_SIZE,
    WORKER_HANDSHAKE_PROTOCOL_VERSION,
    WorkerHandshakeBindingError,
    WorkerHandshakePacketError,
    WorkerHandshakeRequestOptions,
    VerifiedWorkerHandshakeChallenge,
    WorkerTrustConfig,
    _canonical_text,
    _issue_purpose,
    _p256_algorithm,
    _renew_purpose,
    _utc,
    _valid_session_token,
    _verified_packets,
    validate_worker_trust_config,
)
from cap.worker_trust_validate import (
    validate_worker_capability,
    validate_worker_trust_packet,
)


def build_challenge_request(
    config: WorkerTrustConfig,
    options: WorkerHandshakeRequestOptions,
) -> buspacket_pb2.BusPacket:
    """Build, sign, and structurally validate a challenge request."""

    validate_worker_trust_config(config)
    _validate_options(options)
    request = handshake_pb2.WorkerHandshakeChallengeRequest(
        request_id=options.request_id, trace_id=options.trace_id,
        worker_id=config.worker_id, proof_key_id=config.proof_key_id,
        proof_algorithm=_p256_algorithm(), audience=config.audience,
        purpose=options.purpose, client_nonce=bytes(options.client_nonce),
        protocol_version=WORKER_HANDSHAKE_PROTOCOL_VERSION,
        sdk_version=config.sdk_version,
    )
    packet = _new_packet(options.trace_id, config.worker_id, options.created_at)
    packet.worker_handshake_challenge_request.CopyFrom(request)
    sign_trust_packet(packet, config.proof_private_key)
    validate_worker_trust_packet(packet)
    return packet


def build_authenticate(
    config: WorkerTrustConfig,
    verified: VerifiedWorkerHandshakeChallenge,
    capability: handshake_pb2.Handshake,
    current_session_token: str,
    created_at: datetime,
) -> buspacket_pb2.BusPacket:
    """Sign a verified challenge and exact worker capability declaration."""

    validate_worker_trust_config(config)
    challenge = _verified_packets(verified)[1].worker_handshake_challenge
    _bind_challenge_config(config, challenge)
    validate_worker_capability(capability, challenge)
    _validate_session_for_purpose(challenge.purpose, current_session_token)
    packet = _new_packet(challenge.trace_id, config.worker_id, created_at)
    packet.auth_token = current_session_token
    packet.worker_handshake_authenticate.challenge.CopyFrom(challenge)
    packet.worker_handshake_authenticate.capability_handshake.CopyFrom(capability)
    sign_trust_packet(packet, config.proof_private_key)
    validate_worker_trust_packet(packet)
    return packet


def _bind_challenge_config(
    config: WorkerTrustConfig,
    challenge: handshake_pb2.WorkerHandshakeChallenge,
) -> None:
    values = (
        challenge.worker_id == config.worker_id,
        challenge.agent_id == config.expected_agent_id,
        challenge.tenant_id == config.tenant_id,
        challenge.proof_key_id == config.proof_key_id,
        challenge.audience == config.audience,
        challenge.sdk_version == config.sdk_version,
    )
    if not all(values):
        raise WorkerHandshakeBindingError("challenge identity changed")


def _validate_options(options: WorkerHandshakeRequestOptions) -> None:
    if not isinstance(options, WorkerHandshakeRequestOptions):
        raise WorkerHandshakePacketError("request options are required")
    if not _canonical_text(options.request_id) or not _canonical_text(options.trace_id):
        raise WorkerHandshakePacketError("request and trace IDs are invalid")
    if options.purpose not in (_issue_purpose(), _renew_purpose()):
        raise WorkerHandshakePacketError("handshake purpose is unsupported")
    if len(options.client_nonce) != WORKER_HANDSHAKE_NONCE_SIZE:
        raise WorkerHandshakePacketError("client nonce must be 32 bytes")
    _utc(options.created_at)


def _validate_session_for_purpose(purpose: int, token: Optional[str]) -> None:
    if purpose == _issue_purpose() and token:
        raise WorkerHandshakePacketError("issue must not carry a session")
    if purpose == _renew_purpose() and not _valid_session_token(token):
        raise WorkerHandshakePacketError("renew requires the current session")


def _new_packet(
    trace_id: str,
    sender_id: str,
    created_at: datetime,
) -> buspacket_pb2.BusPacket:
    packet = buspacket_pb2.BusPacket(
        trace_id=trace_id, sender_id=sender_id,
        protocol_version=WORKER_HANDSHAKE_PROTOCOL_VERSION,
    )
    try:
        packet.created_at.FromDatetime(_utc(created_at))
    except (ValueError, OverflowError) as exc:
        raise WorkerHandshakePacketError("created_at is invalid") from exc
    return packet


__all__ = ["build_authenticate", "build_challenge_request"]
